### DISCUSSION TITLE
save multiple files opened using mfdataset as a single one

### DISCUSSION THREAD

**Discussion by gkb999** [pydata/xarray#7839] [General] [answered]
Hi all,
I use xarray extensively for my research.
However, for clean netcdf files, I could always manipulate directly without having to save them.
However, my current netcdf files, doesn't have time coordinate I suppose.

It looks like the following: (screenshot attached)
![image](https://github.com/pydata/xarray/assets/7091088/17660228-ea39-4f18-b3f9-376c43841a98)
here the time variable although it is showing as 3280. this is how ds.time looks like:
![image](https://github.com/pydata/xarray/assets/7091088/3af85416-9d3f-45e8-95b1-fea7d884ce91)

For reference: it is a combination of daily data from a particular year (something like say: 2012-2021)
Can I change the time coordinates to the "datetimens" or any standard format that xarray can easily manipulate for further processing?

Also, the x and y cordinates are in 'm', I need them in lat-lon degrees.

Any help on this will be highly appreciated.

**Comment by kmuehlbauer** [comment, marked as answer]
For the time-problem, you would need to add `units` (and `calendar`) attribute. I've tried to prepare the DataArray/Dataset to look like yours.

```python
import xarray as xr
import numpy as np
da = xr.DataArray(np.arange(10), name="time", dims=["time",])
ds = da.to_dataset()
print(ds.time)
ds.time.attrs["units"] = "days since 2012-01-01"
ds.time.attrs["calendar"] = "proleptic_gregorian"
ds = xr.decode_cf(ds)
print(ds)
```
```python
<xarray.DataArray 'time' (time: 10)>
array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
Dimensions without coordinates: time
<xarray.Dataset>
Dimensions:  (time: 10)
Coordinates:
  * time     (time) datetime64[ns] 2012-01-01 2012-01-02 ... 2012-01-10
Data variables:
    *empty*
```

Update: Please also refer to the documentation on [Creating datetime64 data](https://docs.xarray.dev/en/stable/user-guide/time-series.html#creating-datetime64-data).

**Reply by gkb999** [reply]
Awesome, that exactly did what I wanted.
I still think the data doesn't add leap years, but that is still fine, I can ignore those leap years.
How do I add existing variables and the time array?
![image](https://github.com/pydata/xarray/assets/7091088/eebe28ca-4d07-4bbe-ad6a-a54c814b6592)
The data variables are empty. (I'm sorry this could be quite basic, but am new to these kind of analysis)
Many Many thanks in advance already

**Reply by kmuehlbauer** [reply]
Xarray has a very neat documentation which is continuously improved. You'll find answers to these questions in the [user guide](https://docs.xarray.dev/en/stable/user-guide/) and for the specific question in [Data Structures](https://docs.xarray.dev/en/stable/user-guide/data-structures.html)

Here are two versions how this might be achieved to get you on the track:

```python
ds = ds.assign(var1=other_ds.var1)
# or
ds["var1"] = other_ds.var1
```

**Comment by kmuehlbauer** [comment]
For the projection problem we would need more information on the projection. It might be some `polar_stereographic`-projection as this is a variable on the dataset in question. 

If you need the lat/lon just for plotting, this should work if you use the needed parameters for your projection (if it is polar stereographic):

```python
import cartopy.crs as ccrs

map_proj = ccrs.Stereographic(
    true_scale_latitude=60.0, central_latitude=90.0, central_longitude=10.0
)
ds.z.isel(time=0).plot(subplot_kws=dict(projection=map_proj))
```

If you need lat/lon as coordinates (then 2D!) there are several packages out there for coordinate transformation (GDAL, tba.).

**Reply by gkb999** [reply]
This worked too.
Thanks a lot genius :)

**Reply by gkb999** [reply]
> For the projection problem we would need more information on the projection. It might be some `polar_stereographic`-projection as this is a variable on the dataset in question.
> 
> If you need the lat/lon just for plotting, this should work if you use the needed parameters for your projection (if it is polar stereographic):
> 
> ```python
> import cartopy.crs as ccrs
> 
> map_proj = ccrs.Stereographic(
>     true_scale_latitude=60.0, central_latitude=90.0, central_longitude=10.0
> )
> ds.z.isel(time=0).plot(subplot_kws=dict(projection=map_proj))
> ```
> 
> If you need lat/lon as coordinates (then 2D!) there are several packages out there for coordinate transformation (GDAL, tba.).

Just curious why the plot is displayed with a warning:

 `Local\Temp\ipykernel_6472\3827695759.py:3: UserWarning: "true_scale_latitude" parameter is only used for polar stereographic projections. Consider the use of "scale_factor" instead.
  map_proj = ccrs.Stereographic(`.

**Reply by kmuehlbauer** [reply]
Glad it worked. But make sure you use the correct parameters (`true_scale_latitude`, `central_latitude`, `central_longitude`). The example values are for a specific projection used by the German Weather Service.

**Reply by kmuehlbauer** [reply]
> Just curious why the plot is displayed with a warning:
> 
> `Local\Temp\ipykernel_6472\3827695759.py:3: UserWarning: "true_scale_latitude" parameter is only used for polar stereographic projections. Consider the use of "scale_factor" instead. map_proj = ccrs.Stereographic(`.

That's maybe due to using some of my aged examples. Probably cartopy evolved and we have to use/adapt the projection. I'll leave this for you to explore.

**Comment by gkb999** [comment]
Yeah, I changed to south pole, I noticed it is for north.
That's really quick and extremely helpful. :)




*Girija Kalyani BuradaDoctoral StudentSchool of Geography, Environment and
Earth SciencesVictoria University of WellingtonNew Zealand*

**Comment by gkb999** [comment]
Oh yeah, fine :)
This is sufficient .
Thanks a lot again :)





*Girija Kalyani BuradaDoctoral StudentSchool of Geography, Environment and
Earth SciencesVictoria University of WellingtonNew Zealand*
